runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/state/GrpcStateServiceTest.java

```diff
@@ -0,0 +1,133 @@
+import org.apache.beam.model.fnexecution.v1.BeamFnApi;
+import org.apache.beam.sdk.fn.test.TestStreams;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.junit.Assert.assertThat;
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.*;
+
+/** Tests for {@link org.apache.beam.runners.fnexecution.state.GrpcStateService} */
+@RunWith(JUnit4.class)
+public class GrpcStateServiceTest {
+  private static long TIMEOUT = 30 * 1000;
+
+  private GrpcStateService stateService;
+
+  @Mock
+  private StreamObserver<BeamFnApi.StateResponse> responseObserver;
+
+  @Mock
+  private StateRequestHandler handler;
+
+  @Before
+  public void setUp() throws Exception {
+    MockitoAnnotations.initMocks(this);
+    stateService = new GrpcStateService();
+  }
+
+  /**
+   * After a handler has been registered with
+   * {@link GrpcStateService#registerForProcessBundleInstructionId(String, StateRequestHandler)},
+   * the {@link GrpcStateService} should delegate requests through
+   * {@link GrpcStateService#state(StreamObserver)} to the registered handler.
+   */
+  @Test
+  public void testStateRequestsHandledByRegisteredHandlers() throws Exception {
+    // register handler
+    String bundleInstructionId = "bundle_instruction";
+    stateService.registerForProcessBundleInstructionId(bundleInstructionId, handler);
+
+    // open state stream
+    StreamObserver requestObserver = stateService.state(responseObserver);
+
+    // send state request
+    BeamFnApi.StateRequest request =
+        BeamFnApi.StateRequest.newBuilder().setInstructionReference(bundleInstructionId).build();
+    requestObserver.onNext(request);
+
+    // assert behavior
+    verify(handler).accept(eq(request), any());
+  }
+
+  @Test
+  public void testHandlerResponseSentToStateStream() throws Exception {
+    // define handler behavior
+    ByteString expectedResponseData =
+        ByteString.copyFrom("EXPECTED_RESPONSE_DATA", StandardCharsets.UTF_8);
+    String bundleInstructionId = "EXPECTED_BUNDLE_INSTRUCTION_ID";
+    BeamFnApi.StateResponse.Builder expectedBuilder =
+        BeamFnApi.StateResponse
+            .newBuilder()
+            .setGet(BeamFnApi.StateGetResponse.newBuilder().setData(expectedResponseData));
+    StateRequestHandler dummyHandler =
+        (request, result) -> result.toCompletableFuture().complete(expectedBuilder);
+
+    // define observer behavior
+    CompletableFuture<Void> onNextCalled = new CompletableFuture<>();
+    StreamObserver<BeamFnApi.StateResponse> recordingResponseObserver =
+        TestStreams
+            .<BeamFnApi.StateResponse>withOnNext(next -> {
+              synchronized (onNextCalled) {
```

Comment:
Follow the pattern here: https://github.com/apache/beam/blob/cb13361a6d4452e7f5e31af5974e8b90316a654a/sdks/java/harness/src/test/java/org/apache/beam/fn/harness/logging/BeamFnLoggingClientTest.java#L100

Use a concurrent linked queue.

Reply:
Using a BlockingDeque instead in order to poll with a timeout.